x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,361 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.queries;
+
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.TransportVersions;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.action.CoordinatedInferenceAction;
+import org.elasticsearch.xpack.core.ml.action.InferModelAction;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelPrefixStrings;
+import org.elasticsearch.xpack.core.ml.inference.results.TextExpansionResults;
+import org.elasticsearch.xpack.core.ml.inference.results.WarningInferenceResults;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.TextExpansionConfigUpdate;
+import org.elasticsearch.xpack.core.ml.search.TokenPruningConfig;
+import org.elasticsearch.xpack.core.ml.search.WeightedToken;
+import org.elasticsearch.xpack.core.ml.search.WeightedTokensQueryBuilder;
+import org.elasticsearch.xpack.core.ml.search.WeightedTokensUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class SparseVectorQueryBuilder extends AbstractQueryBuilder<SparseVectorQueryBuilder> {
+    public static final String NAME = "sparse_vector";
+    public static final String ALLOWED_FIELD_TYPE = "sparse_vector";
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField INFERENCE_ID_FIELD = new ParseField("inference_id");
+    public static final ParseField TEXT_FIELD = new ParseField("text");
```

Comment:
I think `query_text` (or even `query`) would be better suited here, and pair nicely with `query_vector`.

It's a minor point, but let's align on this as we should have the same for `knn` and `semantic`.

Reply:
++. I vote for using `query_text` for `knn` & `sparse_vector`, since they can (or eventually will) accept vectors or text, and `query` for `semantic` since it only accepts text.

Reply:
I may be missing something, but I don't see us using a `query_text` type of object inside [knn](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-knn-query.html) query? Just `query_vector` which we're already consistent on. 

I am open to changing it to `query` to be consistent with `semantic` query, however I think `query_text` is too verbose. I can also see a very valid argument for keeping it `text` as was originally agreed on as well, because that could potentially keep the door open for alternate objects that are passed in in the future. 

I guess what I'd like to find out, is how strongly do you feel about this naming? I know that we've already closed the first PR pending alignment and discussions, do you feel that `text` needs to be changed?

Reply:
Correct, `knn` query currently doesn't have a text input param (besides indirectly through `query_vector_builder`). This is a forward-looking discussion point :)

I see your point about keeping it `text` to allow for other object types in the future. We could change the `semantic` query to align with this, @carlosdelest WDYT?

Reply:
IMO it's more consistent to use `query_text` as we already have `query_vector` - it's the two sides of the coin, either providing text or the vector for the query. That also keeps the door open for other types (`query_*`) and makes it straightforward that it's related to the actual query vs related to some other config option.

But, you know, _naming_...

> knn query currently doesn't have a text input param

We're planning on adding that, it's just me not having created the JIRA tasks yet :)

Reply:
Resolved internally, will use `query` 👍